### PR TITLE
[Refactor] Refactor train engine

### PR DESCRIPTION
**Pull request by HAOCHENYE**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.12.0) (oldest at bottom):
* __->__ #1518
* #1517
* #1516

Previously, we had two separate train engines:
- `TrainEngine` for regular models
- `VisionComposeTrainEngine` for vision-language models

This duplication led to:
- Code maintenance overhead (242 lines of duplicated logic)
- Tight coupling between engine and model-specific details
- Difficulty in extending to new model types

- **Remove** `VisionComposeTrainEngine` entirely (242 lines deleted)
- **Add** `pre_micro_batch_forward()` and `post_micro_batch_forward()` hooks to `BaseModel`
  - `pre_micro_batch_forward()`: Compute data batch statistics before forward pass
  - `post_micro_batch_forward()`: Aggregate micro-batch results and compute metrics
- **Unify** `TrainEngine` to handle all model types through the new hook system

- **BaseModel**:
  - Add `DataBatchInfo` and `BatchForwardInfo` TypedDicts for return types
  - Implement default `pre_micro_batch_forward()` to compute token statistics
  - Implement default `post_micro_batch_forward()` to aggregate losses and extra info
  - Add overload type hints for `__call__` to improve type inference

- **MoE Model**:
  - Override `post_micro_batch_forward()` to handle MoE-specific logic:
    - Compute maxvio for router load balancing
    - Update router bias based on expert load
  - Add `need_update_bias` property for cleaner code
  - Properly scale balancing_loss and z_loss by batch_size

- **ComposeModel**:
  - Override `pre_micro_batch_forward()` to compute image token statistics
  - Add `ComposeDataBatchInfo` with `step_consumed_img_tokens` field

- **TrainEngine**:
  - Simplify `train_step()` to delegate statistics to model hooks
  - Replace `LossLog` and `OtherLog` with unified `TrainStepInfo`
  - Add `_get_total_loss()` to aggregate all losses (with TODO for future refactor)
  - Remove all model-specific branching logic

- **EngineConfig**:
  - Remove conditional logic for VisionComposeTrainEngine
  - Use single TrainEngine.build() path

- Update to use `TrainStepInfo` instead of separate `LossLog` and `OtherLog`
- Simplify hook signatures (from 2 params to 1)
- Remove conditional engine instantiation logic

- Replace `VisionComposeTrainEngine` imports with `TrainEngine`
- Update test assertions to use new `TrainStepInfo` structure
- Remove TypeAdapter validation for deprecated types

Currently, `TrainEngine._get_total_loss()` aggregates losses by iterating
through ModelOutputs fields. This is pragmatic but not ideal:
- **Pros**: Avoids large-scale changes to model forward() logic
- **Cons**: Engine knows about loss field names (coupling)
- **Future**: Model should return total_loss directly (see TODO comment)

`loss_ctx.batch_size` represents the full gradient accumulation batch size,
not intra_layer_micro_batch. This is correctly set in `CELossContext.build_batches()`
and used for scaling balancing_loss and z_loss.

The pre/post hooks provide clean extension points:
- Subclasses can override to add model-specific logic
- Default implementations in BaseModel handle common cases
- No conditional logic needed in engine layer

1. **Code Reduction**: -242 lines (VisionComposeTrainEngine removed)
2. **Better Separation of Concerns**: Engine focuses on training orchestration, models handle their own statistics
3. **Extensibility**: New model types can override hooks without changing engine
4. **Type Safety**: Unified TrainStepInfo with clear field definitions
5. **Maintainability**: Single engine implementation to maintain

- Loss reduce logic still needs clarification (minor issue, doesn't affect training)
- TODO added for future refactor: move total_loss aggregation to model layer
- All format issues (extra blank lines, class formatting) fixed

### CODE REVIEW

File: xtuner/v1/model/moe/moe.py [hunk 1]
```diff
@@ -313,6 +329,35 @@ def forward(
...
+
+        if self.need_update_bias:
+            assert tokens_per_expert_global_for_bias is not None
+            avg_count_load = tokens_per_expert_global_for_bias.float().mean(1)
+            max_load_i, _ = torch.max(tokens_per_expert_global_for_bias, dim=1)
+
+            maxvio_all_layers = (max_load_i - avg_count_load) / avg_count_load
+            maxvio = maxvio_all_layers.mean()
+
+            self.update_bias(tokens_per_expert_global_for_bias, avg_count_load)  # type: ignore
```
Comment (hhaAndroid) [inline xtuner/v1/model/moe/moe.py]: maxvio 一定要打印

File: xtuner/v1/loss/ce_loss.py [hunk 2]
```diff
@@ -146,6 +146,7 @@ def build_batches(  # type: ignore[override]
             dist.all_reduce(global_denominator, op=dist.ReduceOp.SUM)
 
         for loss_ctx in loss_ctx_list:
+            loss_ctx._batch_size = len(loss_ctx_list)
```
Comment (hhaAndroid) [inline xtuner/v1/loss/ce_loss.py:149]: 需要考虑 intra_layer_micro_batch 情况下逻辑
